### ISSUE TITLE
[java] UnusedNullCheck is not aware of Arrays.equals or Objects.equals

### ISSUE THREAD

**Issue by jsotuyod** [pmd/pmd#4907]
**Affects PMD Version:** 
7.0.0

**Rule:**

[`UnusedNullCheckInEquals`](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#unusednullcheckinequals)

**Description:**

The rule is only checking against invocations to `Object.equals`.

**Code Sample demonstrating the issue:**

```java
class NewEqualsHelpers {
    private Object[] multiple;
    private Object single;

    void test(NewEqualsHelpers other){
        if (other.multiple != null && java.util.Arrays.equals(other.multiple, multiple)) { } // should flag

        if (single != null && Objects.equals(single, other.single)) { } // should flag
    }
}
```

**Expected outcome:**

PMD should report a violation at lines 6 and 8.

**Running PMD through:** Any

**Comment by zbynek** [comment]
I assume for
```
if (single != null && Objects.equals(single, other.single))  { }
```
the "correct" code is
```
if (single != null && single.equals(other.single)) { }
```
as it's both shorter and avoids checking for null twice. However for
```
if (other.multiple != null && java.util.Arrays.equals(other.multiple, multiple)) { } 
```
it's not clear what would the correct equivalent be, is that really a false negative?

**Comment by adangel** [comment]
For

```java
// check "single != null" is not needed
if (single != null && Objects.equals(single, other.single))  { }

// check "other.single != null" is not needed
if (other.single != null && Objects.equals(single, other.single))  { }

```

I'd suggest as correct code

```java
if (Objects.equals(single, other.single))  { }
```

I guess, it depends on a personal preference... whether we would suggest using "Objects.equals"..

However, for a really correct code, we are missing a null check for "other"... it could be null. But that's not part of this rule. The point is, the check "single != null" is not needed.
And it's the same for the other check ("other.single != null").

The two false-negative cases for the Arrays are:

```java
// check "other.multiple != null" is not needed
if (other.multiple != null && Arrays.equals(other.multiple, multiple)) { }
// check "multiple != null" is not needed
if (multiple != null && Arrays.equals(other.multiple, multiple)) { }
```

The correct code (using Arrays):
```java
if (Arrays.equals(other.multiple, multiple)) { }
```


Note: the code samples combine the null check with the call to Objects.equals/Arrays.equals. That means, there is no special handling if one of the two values would be null - they are just not considered to be equal. E.g. we should not flag something like:

```java
if (single != null && Objects.equals(single, other.single))  {
  // single and other.single are _not_ null for sure and are equal
} else {
  // either single or other.single is null or both are null
  // or both are given but are _not_ equal
}
```

What happens, if both values would be null? -> that could be a problem if we suggest to only use "Objects.equals". Arrays.equals and Objects.equals both return true, if both values are null. So, suggesting these methods without a warning could be surprising...

**Comment by zbynek** [comment]
> I'd suggest as correct code
> `if (Objects.equals(single, other.single))  { }`

as you mention later on, that does not preserve the functionality if `single` is `null`. I think the main reason for using `Objects.equals(a,b)` instead of just `a.equals(b)` is null safety, so I don't think the rule should suggest ignoring the case when `single` is null.

**Comment by adangel** [comment]
So, yes, that means, that neither case is actually a false negative.

Shall we just close it as "not planned" (maybe labelling it with "was:invalid")?

**Comment by zbynek** [comment]
Yes, let's close this.
